aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -7,73 +7,252 @@
 #include <c10/core/TensorOptions.h>
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <ATen/core/Dimname.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_specialization_of
+//
+template <template <typename...> class T, typename U>
+struct is_specialization_of: std::false_type {};
+
+template <template <typename...> class T, typename... Us>
+struct is_specialization_of<T, T<Us...>>: std::true_type {};
+
+// is_tuple_of_lvalue_refs
+//
+template<class T, bool = is_specialization_of<std::tuple, T>::value>
+struct is_tuple_of_lvalue_refs :
+  guts::typelist::all<std::is_lvalue_reference, guts::typelist::from_tuple_t<T>>
+{};
+
+template<class T>
+struct is_tuple_of_lvalue_refs<T, false> : std::false_type {};
+
+// tuple_elements
+//
+template <class Tuple, size_t... ns>
```

Comment:
`tuple_elements` and `tuple_take` seem generic enough to warrant being somewhere more central, maybe `c10/util/Metaprogramming.h`, and get some unit tests in `Metaprogramming_test.cpp`.

Reply:
agree! done